libraries/src/AWS.Lambda.Powertools.Logging/Internal/PowertoolsLogger.cs

```diff
@@ -228,24 +251,120 @@ public bool IsEnabled(LogLevel logLevel)
-            CustomFormatter(state, exception, out var customMessage) && customMessage is not null
-                ? customMessage
-                : formatter(state, exception));
+        logEntry.TryAdd(LoggingConstants.KeyTimestamp, timestamp.ToString("o"));
+        logEntry.TryAdd(LoggingConstants.KeyLogLevel, logLevel.ToString());
+        logEntry.TryAdd(LoggingConstants.KeyService, Service);
+        logEntry.TryAdd(LoggingConstants.KeyLoggerName, _name);
+        logEntry.TryAdd(LoggingConstants.KeyMessage, message);
+
         if (CurrentConfig.SamplingRate.HasValue)
-            message.TryAdd(LoggingConstants.KeySamplingRate, CurrentConfig.SamplingRate.Value);
+            logEntry.TryAdd(LoggingConstants.KeySamplingRate, CurrentConfig.SamplingRate.Value);
         if (exception != null)
-            message.TryAdd(LoggingConstants.KeyException, exception);
+            logEntry.TryAdd(LoggingConstants.KeyException, exception);
+
+        return logEntry;
+    }
+
+    /// <summary>
+    ///     Gets a formatted log entry.
+    /// </summary>
+    /// <param name="logLevel">Entry will be written on this level.</param>
+    /// <param name="timestamp">Entry timestamp.</param>
+    /// <param name="message">The message to be written. Can be also an object.</param>
+    /// <param name="exception">The exception related to this entry.</param>
+    /// <param name="logFormatter">The custom log entry formatter.</param>
+    private object GetFormattedLogEntry(LogLevel logLevel, DateTime timestamp, object message,
+        Exception exception, ILogFormatter logFormatter)
+    {
+        if (logFormatter is null)
+            return null;
+
+        var logEntry = new LogEntry
+        {
+            Timestamp = timestamp,
+            Level = logLevel,
+            Service = Service,
+            Name = _name,
+            Message = message,
+            Exception = exception,
+            SamplingRate = CurrentConfig.SamplingRate,
+        };
+
+        var extraKeys = new Dictionary<string, object>();
+
+        // Add Custom Keys
+        foreach (var (key, value) in Logger.GetAllKeys())
+        {
+            switch (key)
+            {
+                case LoggingConstants.KeyColdStart:
+                    logEntry.ColdStart = (bool)value;
+                    break;
+                case LoggingConstants.KeyXRayTraceId:
+                    logEntry.XRayTraceId = value as string;
+                    break;
+                case LoggingConstants.KeyCorrelationId:
+                    logEntry.CorrelationId = value as string;
+                    break;
+                default:
+                    extraKeys.TryAdd(key, value);
+                    break;
+            }
+        }
+
+        // Add Extra Fields
+        if (CurrentScope?.ExtraKeys is not null)
+        {
+            foreach (var (key, value) in CurrentScope.ExtraKeys)
+            {
+                if (!string.IsNullOrWhiteSpace(key))
```

Comment:
don't think is possible at runtime to have an empty or null key for dictionary, this could be simplified with `logEntry.ExtraKeys = new Dictionary<string, string>(extraKeys);`

Reply:
It cannot, I can remove the null check but is's an append to the existing dictionary, so cannot be replaced.

Reply:
I have replaced it with a new throw new LogFormatException (custom exception)